acme/acme/client.py

```diff
@@ -549,14 +550,82 @@ def new_account(self, new_account):
 
         :returns: Registration Resource.
         :rtype: `.RegistrationResource`
-
         """
         response = self.net.post(self.directory['newAccount'], new_account,
             acme_version=2)
         # "Instance of 'Field' has no key/contact member" bug:
         # pylint: disable=no-member
         return self._regr_from_response(response)
 
+    def new_order(self, csr_pem):
```

Comment:
Any reason to require the CSR now rather than allowing the user to provide it later? All we need now is the list of domains.

Reply:
An earlier ACME draft specified the CSR up front rather than at the end. We've changed the spec, but I figured I'd leave the CSR argument at the beginning of the process, with the idea that that gives us more flexibility under the hood. I figured changing the flow to take the CSR at the end would have been sort of overfitting the current protocol rather than abstracting from it.

Reply:
> An earlier ACME draft specified the CSR up front rather than at the end.

Yeah I actually chimed in on the acme list in favor of that approach, but I don't think it will matter much for us in practice.

> We've changed the spec, but I figured I'd leave the CSR argument at the beginning of the process, with the idea that that gives us more flexibility under the hood. I figured changing the flow to take the CSR at the end would have been sort of overfitting the current protocol rather than abstracting from it.

What flexibility does this buy us? I might be missing something you thought about. To be honest, I don't have a strong preference here, but I think only requiring what you actually need simplifies the function and makes it more flexible for the user.

Reply:
Talked about this offline; I think my preference for up-front was a bit stronger than Brad's preference for not, so keeping it this way.